File: ios/RCTNendMediaViewManager.m
//
//  RCTNendMediaViewManager.m
//  AwesomeProject
//

#import "RCTNendMediaViewManager.h"

#import <React/RCTView.h>
#import <React/RCTUIManager.h>

#import "RCTNendVideoNativeAd.h"

@import NendAd;

@interface RCTMediaView : RCTView <NADNativeVideoViewDelegate>

@property (nonatomic, strong) NADNativeVideoView *videoView;
@property (nonatomic, copy) RCTBubblingEventBlock onPlaybackStarted;
@property (nonatomic, copy) RCTBubblingEventBlock onPlaybackStopped;
@property (nonatomic, copy) RCTBubblingEventBlock onPlaybackCompleted;
@property (nonatomic, copy) RCTBubblingEventBlock onPlaybackError;
@property (nonatomic, copy) RCTBubblingEventBlock onFullScreenOpened;
@property (nonatomic, copy) RCTBubblingEventBlock onFullScreenClosed;

@end

@implementation RCTMediaView

- (instancetype)init
{
  self = [super init];
  if (self) {
    _videoView = [[NADNativeVideoView alloc] initWithFrame:CGRectZero];
    _videoView.translatesAutoresizingMaskIntoConstraints = NO;
    _videoView.delegate = self;
    [self addSubview:_videoView];
    [NSLayoutConstraint activateConstraints:@[[_videoView.leftAnchor constraintEqualToAnchor:self.leftAnchor],
                                              [_videoView.topAnchor constraintEqualToAnchor:self.topAnchor],
                                              [_videoView.rightAnchor constraintEqualToAnchor:self.rightAnchor],
                                              [_videoView.bottomAnchor constraintEqualToAnchor:self.bottomAnchor]]];
  }
  return self;
}

#pragma mark - NADNativeVideoViewDelegate

- (void)nadNativeVideoViewDidStartPlay:(NADNativeVideoView *)videoView
{
  if (self.onPlaybackStarted) {
    self.onPlaybackStarted(nil);
  }
}

- (void)nadNativeVideoViewDidStopPlay:(NADNativeVideoView *)videoView
{
  if (self.onPlaybackStopped) {
    self.onPlaybackStopped(nil);
  }
}

- (void)nadNativeVideoViewDidCompletePlay:(NADNativeVideoView *)videoView
{
  if (self.onPlaybackCompleted) {
    self.onPlaybackCompleted(nil);
  }
}

- (void)nadNativeVideoViewDidFailToPlay:(NADNativeVideoView *)videoView
{
  if (self.onPlaybackError) {
    self.onPlaybackError(nil);
  }
}

- (void)nadNativeVideoViewDidOpenFullScreen:(NADNativeVideoView *)videoView
{
  if (self.onFullScreenOpened) {
    self.onFullScreenOpened(nil);
  }
}

- (void)nadNativeVideoViewDidCloseFullScreen:(NADNativeVideoView *)videoView
{
  if (self.onFullScreenClosed) {
    self.onFullScreenClosed(nil);
  }
}

@end

#pragma mark -

@implementation RCTNendMediaViewManager

RCT_EXPORT_MODULE()

- (UIView *)view
{
  return [RCTMediaView new];
}

RCT_EXPORT_METHOD(setMedia:(nonnull NSNumber *)reactTag refId:(nonnull NSNumber *)refId)
{
  RCTNendVideoNativeAd *module = [self.bridge moduleForClass:[RCTNendVideoNativeAd class]];
  [self.bridge.uiManager addUIBlock:^(RCTUIManager *uiManager, NSDictionary<NSNumber *,UIView *> *viewRegistry) {
    RCTMediaView *view = (RCTMediaView *)viewRegistry[reactTag];
    if (view && module) {
      NADNativeVideo *videoAd = [module getVideoAdWithRefernceId:refId.integerValue];
      if (videoAd) {
        view.videoView.videoAd = videoAd;
      }
    }
  }];
}

RCT_EXPORT_VIEW_PROPERTY(onPlaybackStarted, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onPlaybackStopped, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onPlaybackCompleted, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onPlaybackError, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onFullScreenOpened, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onFullScreenClosed, RCTBubblingEventBlock)

@end

File: ios/RCTNendMediaViewManager.swift
//
//  RCTNendMediaViewManager.swift
//  AwesomeProject
//

import UIKit
import NendAd

@objc(RCTNendMediaViewManager)
class RCTNendMediaViewManager: RCTViewManager {
  static override func requiresMainQueueSetup() -> Bool {
    return false
  }
  
  override var methodQueue: DispatchQueue! {
    return DispatchQueue.main
  }

  override func view() -> UIView! {
    return RCTMediaView()
  }
  
  @objc(setMedia:refId:)
  func setMedia(reactTag: NSNumber, refId: Int) {
    bridge.uiManager.addUIBlock { (manager, viewRegistry) in
      if let view = viewRegistry?[reactTag] as? RCTMediaView,
        let module = self.bridge.module(for: RCTNendVideoNativeAd.self) as? RCTNendVideoNativeAd,
        let videoAd = module.getVideoAd(refernceId: refId) {
        view.videoView.videoAd = videoAd
      }
    }
  }
}

class RCTMediaView: RCTView {
  var onPlaybackStarted: RCTBubblingEventBlock?
  var onPlaybackStopped: RCTBubblingEventBlock?
  var onPlaybackCompleted: RCTBubblingEventBlock?
  var onPlaybackError: RCTBubblingEventBlock?
  var onFullScreenOpened: RCTBubblingEventBlock?
  var onFullScreenClosed: RCTBubblingEventBlock?
  
  fileprivate var videoView: NADNativeVideoView!
  
  override init(frame: CGRect) {
    super.init(frame: frame)
    createVideoView(frame: frame)
  }
  
  required init?(coder aDecoder: NSCoder) {
    super.init(coder: aDecoder)
    createVideoView()
  }
  
  private func createVideoView(frame: CGRect = CGRect(x: 0, y: 0, width: 0, height: 0)) {
    videoView = NADNativeVideoView(frame: frame)
    videoView.translatesAutoresizingMaskIntoConstraints = false
    videoView.delegate = self
    self.addSubview(videoView)
    NSLayoutConstraint.activate([
      videoView.leftAnchor.constraint(equalTo: self.leftAnchor),
      videoView.topAnchor.constraint(equalTo: self.topAnchor),
      videoView.rightAnchor.constraint(equalTo: self.rightAnchor),
      videoView.bottomAnchor.constraint(equalTo: self.bottomAnchor)])
  }
}

extension RCTMediaView: NADNativeVideoViewDelegate {
  func nadNativeVideoViewDidStopPlay(_ videoView: NADNativeVideoView) {
    if let callback = onPlaybackStopped {
      callback(nil)
    }
  }
  
  func nadNativeVideoViewDidFail(toPlay videoView: NADNativeVideoView) {
    if let callback = onPlaybackError {
      callback(nil)
    }
  }
  
  func nadNativeVideoViewDidStartPlay(_ videoView: NADNativeVideoView) {
    if let callback = onPlaybackStarted {
      callback(nil)
    }
  }
  
  func nadNativeVideoViewDidCompletePlay(_ videoView: NADNativeVideoView) {
    if let callback = onPlaybackCompleted {
      callback(nil)
    }
  }
  
  func nadNativeVideoViewDidOpenFullScreen(_ videoView: NADNativeVideoView) {
    if let callback = onFullScreenOpened {
      callback(nil)
    }
  }
  
  func nadNativeVideoViewDidCloseFullScreen(_ videoView: NADNativeVideoView) {
    if let callback = onFullScreenClosed {
      callback(nil)
    }
  }
}
